File: src/main/java/de/steklopod/sec09/Lec06Assignment.java
package de.steklopod.sec09;

import de.steklopod.courseutil.Util;
import de.steklopod.sec09.assignment.OrderProcessor;
import de.steklopod.sec09.assignment.OrderService;
import de.steklopod.sec09.assignment.PurchaseOrder;
import reactor.core.publisher.Flux;

import java.util.Map;
import java.util.Set;
import java.util.function.Function;

public class Lec06Assignment {

    public static void main(String[] args) {

        Map<String, Function<Flux<PurchaseOrder>, Flux<PurchaseOrder>>> map = Map.of(
                "Kids", OrderProcessor.kidsProcessing(),
                "Automotive", OrderProcessor.automotiveProcessing()
        );

        Set<String> set = map.keySet();

        OrderService.orderStream()
                .filter(p -> set.contains(p.getCategory()))
                .groupBy(PurchaseOrder::getCategory)  // 2 keys
                .flatMap(gf -> map.get(gf.key()).apply(gf)) //flux
                .subscribe(Util.subscriber());

        Util.sleepSeconds(60);

    }


}

File: src/main/kotlin/de/steklopod/sec09/Lec06Assignment.kt
package de.steklopod.sec09

import de.steklopod.sec09.assignment.OrderProcessor
import de.steklopod.sec09.assignment.OrderService
import de.steklopod.sec09.assignment.PurchaseOrder
import de.steklopod.utils.Util.sleepSeconds
import de.steklopod.utils.Util.subscriber
import reactor.core.publisher.GroupedFlux
import java.util.Map

object Lec06Assignment {
    @JvmStatic
    fun main(args: Array<String>) {
        val map = Map.of(
            "Kids", OrderProcessor.kidsProcessing(),
            "Automotive", OrderProcessor.automotiveProcessing()
        )
        val set: Set<String> = map.keys
        OrderService.orderStream()
            .filter { p: PurchaseOrder -> set.contains(p.category) }
            .groupBy { obj: PurchaseOrder -> obj.category } // 2 keys
            .flatMap { gf: GroupedFlux<String, PurchaseOrder> ->
                map[gf.key()]!!
                    .apply(gf)
            } //flux
            .subscribe(subscriber())
        sleepSeconds(60)
    }
}
